kernel/arch/dreamcast/kernel/cache.s

```diff
@@ -74,99 +69,178 @@ flush_loop:
 	rts
 	nop
 
-	.align	2
-ormask:
-	.long	0x100000f0
-addrarray:
-	.long	0xf0000000	! CACHE_IC_ADDRESS_ARRAY
-entrymask:
-	.long	0x1fe0		! CACHE_IC_ENTRY_MASK
-validmask:
-	.long	0xfffffc00
-	
 
-! Goes through and invalidates the O-cache for a given block of
-! RAM. Make sure that you've called dcache_flush_range first if
-! you care about the contents.
+! This routine goes through and invalidates the dcache for a given 
+! range of RAM. Make sure that you've called dcache_flush_range first
+! if you care about the contents.
+!
 ! r4 is starting address
 ! r5 is count
 _dcache_inval_range:
 	! Get ending address from count and align start address
-	add	r4,r5
-	mov.l	l1align,r0
-	and	r0,r4
+	add	    r4, r5
+	mov.l	align_mask, r0
+	and	    r0, r4
 
-dinval_loop:
-	! Invalidate the O cache
+.dinval_loop:
+	! Invalidate the dcache
 	ocbi	@r4
-	cmp/hs	r4,r5
-	bt/s	dinval_loop
-	add	#32,r4		! += CPU_CACHE_BLOCK_SIZE
+	cmp/hs	r4, r5
+	bt/s	.dinval_loop
+	add	    #32, r4		! Move on to next cache block
 
 	rts
 	nop
 
 
-! This routine just goes through and forces a write-back on the
+! This routine goes through and forces a write-back on the
 ! specified data range. Use prior to dcache_inval_range if you
-! care about the contents.
+! care about the contents. If the range is bigger than the dcache,
+! we flush the whole cache instead.
+!
 ! r4 is starting address
 ! r5 is count
 _dcache_flush_range:
-	! Get ending address from count and align start address
-	add	r4,r5
-	mov.l	l1align,r0
-	and	r0,r4
+	! Divide byte count by 32 
+	mov     #-5, r1
+	shad    r1, r5           
+
+	! Compare with 512
+	mov.w   cache_lines, r2
+	cmp/hi  r2, r5
+	bt      _dcache_flush_all  ! If lines > 512, jump to _dcache_flush_all
```

Comment:
What if you flush buffer that almost not in cache with size like 515 cache lines. So a flush_all still will be faster? I mean, maybe a gap is needed?

Reply:
Yes, you are correct! I just tested flushing 16kb range and 16kb range +32. There is a huge time difference(60,000 ns) that occurs when we start relying on the hard coded 512 check. I can up the check(# >512) and find the sweet spot. For flushing typical ranges like 128*128*2, 256*256*2 or 512*512*2, dcache_flush_all() would be faster.